### ISSUE TITLE
Use custom data provider for nested document

### ISSUE THREAD

**Issue by pierredup** [api-platform/core#1214]
As far as I can tell, when using nested documents you currently must use an IRI. But I would like to use a string value for the nested object, and use a custom data provider to retrieve the proper object.

For example, let's say you have a Category and a Product entity. If I want to add a product to a category, I have to use an IRI:

```json
// POST /api/products

{
    "category": "/api/categories/123",
    "name": "product-name"
    ....
}
```

What I would like is to have the following:

```json
// POST /api/products
{
    "category": "category-slug",
    "name": "product-name"
    ....
}
```

Then using a custom data provider, lookup the value of `category-slug` myself and return the full object.

I'm not sure if this is currently possible, but I tried many different configs without luck. Is it something that is supported? If not, would it be accepted if I create a PR to add such functionality?

**Comment by soyuka** [comment]
Indeed, this isn't really valid JSON-LD and you may want to implement a custom denormalizer to handle such case. I'm not sure that we'd accept a generic workaround for this as it's not following the specs.

**Comment by pierredup** [comment]
It doesn't work with a custom normalizer, I have tried but the failure keeps happening before my normalizer is hit because [this condition](https://github.com/api-platform/core/blob/a24d2bbe9331d5b86d48b13043278138ab30e386/src/Serializer/AbstractItemNormalizer.php#L294) always returns false

**Comment by soyuka** [comment]
If I understand correctly you're trying to do something like this?

```
public function supportsDenormalization($data, $type, $format = null)
{
  return $data instanceof Product;
}

public function denormalize($data, $class, $format = null, array $context = [])
{      
    //test things before this maybe
    $data['category'] = $this->categoryRepository->findOneBySlug($data['category']);

    return parent::denormalize($data, $class, $format, $context);
}

```

**Comment by pierredup** [comment]
Something like that, yes

**Comment by pierredup** [comment]
Yes, I just tested with the latest version, and I still have the same issue.

From what I can gather, [this statement](https://github.com/api-platform/core/blob/8fcc41592617ffff2cbd1b7db7ed557e0bc603e8/src/Serializer/AbstractItemNormalizer.php#L304-L307) always return `false`.
I added the `@ApiProperty(writableLink=true)` annotation to the property, but in the linked code, `$propertyMetadata->isWritableLink()` always returns `false`.
I haven't had time yet to figure out why the annotation doesn't set the metadata correctly, but if I add ` || true` to that condition (so basically just force it inside the `if`), then it works as expected with a custom denormalizer
